pkg/hydrator/batch.go

```diff
@@ -0,0 +1,155 @@
+	VulnerabilitySolutionsFetcher VulnerabilitySolutionsFetcher
+	BatchSolutionFetcher          BatchSolutionFetcher
+}
+
+func (a *assetVulnerabilityHydrator) HydrateAssetVulnerability(ctx context.Context, assetVulnerability nexposeAssetVulnerability) (domain.VulnerabilityDetails, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	errorChan := make(chan error, 2)
+
+	vulnerabilityDetailsChan := make(chan nexposeVulnerability, 1)
+	go func() {
+		details, err := a.VulnerabilityDetailsFetcher.FetchVulnerabilityDetails(ctx, assetVulnerability.ID)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		vulnerabilityDetailsChan <- details
+	}()
+
+	solutionsChan := make(chan []string, 1)
+	go func() {
+		solutionIDs, err := a.VulnerabilitySolutionsFetcher.FetchVulnerabilitySolutions(ctx, assetVulnerability.ID)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		solutions, err := a.BatchSolutionFetcher.BatchFetchSolution(ctx, solutionIDs)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		solutionsChan <- solutions
+	}()
+
+	vulnerabilityDetails := domain.VulnerabilityDetails{
+		ID:      assetVulnerability.ID,
+		Results: assetVulnerability.Results,
+	}
+
+	for i := 0; i < 2; i = i + 1 {
+		select {
+		case solutions := <-solutionsChan:
+			vulnerabilityDetails.Solutions = solutions
+		case nexposeVulnDetails := <-vulnerabilityDetailsChan:
+			vulnerabilityDetails.CvssV2Score = nexposeVulnDetails.CvssV2Score
+			vulnerabilityDetails.CvssV2Severity = nexposeVulnDetails.CvssV2Severity
+			vulnerabilityDetails.Description = nexposeVulnDetails.Description
+			vulnerabilityDetails.Title = nexposeVulnDetails.Title
+		case err := <-errorChan:
+			cancel()
+			return domain.VulnerabilityDetails{}, err
+		}
+	}
+
+	return vulnerabilityDetails, nil
+}
+
+// BatchAssetVulnerabilityHydrator represents an interface for concurrently hydrating asset vulnerabilities
+type BatchAssetVulnerabilityHydrator interface {
+	BatchHydrateAssetVulnerabilities(ctx context.Context, assetVulns []nexposeAssetVulnerability) ([]domain.VulnerabilityDetails, error)
+}
+
+type batchAssetVulnerabilityHydrator struct {
+	AssetVulnerabilityHydrator AssetVulnerabilityHydrator
+}
+
+func (b *batchAssetVulnerabilityHydrator) BatchHydrateAssetVulnerabilities(ctx context.Context, assetVulns []nexposeAssetVulnerability) ([]domain.VulnerabilityDetails, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	vulnerabilityDetailsChan := make(chan domain.VulnerabilityDetails, len(assetVulns))
+	errorChan := make(chan error, 1)
+
+	for _, assetVuln := range assetVulns {
+		go func(assetVuln nexposeAssetVulnerability) {
+			details, err := b.AssetVulnerabilityHydrator.HydrateAssetVulnerability(ctx, assetVuln)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+			vulnerabilityDetailsChan <- details
+		}(assetVuln)
+	}
+
+	vulnerabilityDetails := make([]domain.VulnerabilityDetails, len(assetVulns))
```

Comment:
task: This needs to change to using `make(thing, 0, len(assetVulns))` and `append`.